### PR TITLE
Projects: prevent report abuse spam

### PR DESCRIPTION
**Pull request by Erin007**
[LP-52](https://codedotorg.atlassian.net/browse/LP-52)

Kids are overzealous with the report abuse feature; they will often report abuse on the same project multiple times when the project isn't abusive, which is a pain because it floods Zendesk with tickets. To help alleviate this issue, I'm preventing users from re-reporting abuse on projects.  If a user has already reported a project, the report abuse link will be hidden on that project for that user. To do this, each time a project is reported for abuse, I store the project's channel id in a cookie `reported_abuse` for the user. 
Screen Shot 2019-05-30 at 3 15 22 PM

There are 3 places where this change will affect: 

1.) The footer menu on a project view page for App Lab 
Screen Shot 2019-05-24 at 12 22 29 PM

2.) The footer menu on a project view page for all other project types 
Screen Shot 2019-05-24 at 4 29 51 PM

3.) The mini footer menu on a project edit page for all project types
Screen Shot 2019-05-30 at 3 25 41 PM

While working on this, I found that we generate the contents of the footer in 3 separate places. We should investigate if these can be consolidated: [STAR-464](https://codedotorg.atlassian.net/browse/STAR-464)

Also, I'm intentionally not handing the report abuse link in the admin panel.  If you're an admin, you're internal and we trust you not to excessively spam Zendesk. 🙂

### CODE REVIEW

**Comment by jmkulwik** [comment]
Planning to add any tests?

**Comment by davidsbailey** [comment]
Hi Erin, this approach looks like it will work. :-) however, I want to point out that the cookie name will collide across environments, so if you report abuse on a certain project on staging, you won't be able to report the same project in prod. this should be fine, because project ids are unlikely to collide, so this is just a warning to think twice before copying this pattern in the future.

For next time, it's worth considering using the rails session, which has the advantages of (1) separate storage for localhost / staging / prod environments, and (2) sticks with users across different browsers while still also tracking signed-out users. 

the main limitation of the rails session is that you cannot (to my knowledge) modify it from the client. you could work around this here, but since there's no compelling reason I'm not requesting that you reimplement this that way.

**Comment by davidsbailey** [comment]
here is one example where we use the rails session: https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/lib/version_redirect_overrider.rb

File: apps/src/applab/applab.js [hunk 1]
```diff
@@ -200,6 +203,20 @@ Applab.makeFooterMenuItems = function(isIframeEmbed) {
...
+  var userAlreadyReportedAbuse =
+    cookies.get('reported_abuse') &&
+    _.includes(
+      JSON.parse(cookies.get('reported_abuse')),
+      project.getCurrentId()
+    );
+
+  if (userAlreadyReportedAbuse) {
+    _.remove(footerMenuItems, function(menuItem) {
+      return menuItem.key === 'reportAbuse';
```
Comment (davidsbailey) [inline apps/src/applab/applab.js]: this is spelled `'report-abuse'` in 2 places below. is this a typo here? please make sure this is working as expected in applab.

**Review by davidsbailey — APPROVED** [review]
Nice implementation, @Erin007 !

**Comment by Erin007** [comment]
> however, I want to point out that the cookie name will collide across environments, so if you report abuse on a certain project on staging, you won't be able to report the same project in prod. this should be fine, because project ids are unlikely to collide, so this is just a warning to think twice before copying this pattern in the future.

Agreed that project ids likely won't collide, but seems like best practice to use this since we have it 🙂: https://github.com/code-dot-org/code-dot-org/blob/8990f16a29d8a1cf9dba0ffb220bd424e8542810/dashboard/lib/cookie_helpers.rb#L7

I'll update

**Comment by Erin007** [comment]
> > however, I want to point out that the cookie name will collide across environments, so if you report abuse on a certain project on staging, you won't be able to report the same project in prod. this should be fine, because project ids are unlikely to collide, so this is just a warning to think twice before copying this pattern in the future.
> 
> Agreed that project ids likely won't collide, but seems like best practice to use this since we have it 🙂:
> 
> https://github.com/code-dot-org/code-dot-org/blob/8990f16a29d8a1cf9dba0ffb220bd424e8542810/dashboard/lib/cookie_helpers.rb#L7
> 
> I'll update

On second thought that' a ruby helper available available on dashboard. It doesn't seem important enough to thread through into the JS files since the project ids won't overlap.

**Comment by breville** [comment]
Do we cap how big this cookie can get?

**Comment by Erin007** [comment]
> Do we cap how big this cookie can get?

Not right now. Looks like browsers can support up to [4093 bytes of cookie](https://www.lifewire.com/cookie-size-limit-3466810). There are 43 characters in a project id. Even with all of our other cookies, I don't think we'd be getting close to too big even if a student did go on a reporting rampage. We could check the length?

**Comment by breville** [comment]
I'm not sure how cautious we want to be.  It seems it would be unlikely to overflow under normal usage, but I know that we do see `CookieOverflow` errors in honeybadger for other cookies sometimes, and I am curious what happens if a particularly active user does overflow their cookie... does it break anything for them or us?

**Comment by Erin007** [comment]
> Planning to add any tests?

@jmkulwik I added UI tests for App Lab and non-App Lab projects for both the mini footer on the view page and the footer on the share page. Is that what you had in mind?

**Comment by Erin007** [comment]
> I'm not sure how cautious we want to be. It seems it would be unlikely to overflow under normal usage, but I know that we do see `CookieOverflow` errors in honeybadger for other cookies sometimes, and I am curious what happens if a particularly active user does overflow their cookie... does it break anything for them or us?

I did I little experiment. 

If I do this, the `reported_abuse` cookie never gets set.
```
    const length = 4094
    const string = "x"
    const extraLongString = string.repeat(length)
    cookies.set('reported_abuse', extraLongString)
```

If the length is smaller, then then cookie is set. 

In either scenario, the submit of the report abuse form goes through just fine and the project is blocked. So, in the unlikely event someone's gone on a long enough reporting rampage to have a massive cookie the only thing I think will "break" is our ability to track which projects they've reported abuse on. I'm inclined to cross this bridge if we come to it. 

If it becomes a problem, I imagine we might set some threshold for number of projects a single user has reported, check the length of the cookie to see if we've hit that threshold and then say "hey, we take abuse really seriously and it looks like you're just messing around." 🤷‍♀️Hopefully, this change makes Josh's life easier in the meantime.

**Comment by breville** [comment]
@Erin007 Love the investigation.  This seems good for now.  (As a side note, I wonder how we got overflowing cookies?  Maybe those are `HttpOnly` cookies?)

**Comment by Erin007** [comment]
@breville Maybe `HttPOnly` cookies? I also wonder if it's the cumulative size of all of the cookies together that makes them overflow? It looks like I can't set one cookie's value over the storage limit, but that doesn't mean the combined total of many cookies won't be too big.

**Comment by davidsbailey** [comment]
LGTM! good investigation, I don't think cookie overflow concerns should block merging.

**Pull request merged**